polynote-kernel/src/main/scala/polynote/kernel/lang/scal/ScalaSource.scala

```diff
@@ -83,9 +83,11 @@ class ScalaSource[Interpreter <: ScalaInterpreter](val interpreter: Interpreter)
 
   // the trees that came out of the parser, ignoring any parse errors
   private lazy val parsedTrees = parsed match {
-    case Ior.Both(_, trees) => Right(trees)
-    case Ior.Right(trees) => Right(trees)
-    case Ior.Left(err) => Left(err)
+    case Ior.Both(_, trees) if trees.nonEmpty => Right(trees)
+    case Ior.Both(_, _)                       => Left(EmptyCell)
+    case Ior.Right(trees) if trees.nonEmpty   => Right(trees)
+    case Ior.Right(_)                         => Left(EmptyCell)
+    case Ior.Left(err)                        => Left(err)
```

Comment:
If we are ignoring parse errors (this is used mainly for completions?) shouldn't we have 
```
case Ior.Left(err)                        => Left(EmptyCell)
``` 
instead?

Reply:
This is kind of confusing as it is – I might change it to be `Ior` all the way through at some point, rather then schlepping to various `Either`s that have different meanings of `Left`.

In this case,
* `Right` means there was no parse error, so it contains trees. If those are empty, fail fast, otherwise continue (i.e. if the parse error was "recoverable" we can still do completions)
* `Both` means there was a parse error, but we still got trees. If those are empty, fail fast, otherwise continue
* `Left` means there was an actual exception during parsing, so we can't do anything else because we don't have trees. We want to surface that exception, though, which is why it doesn't get replaced with `EmptyCell`. The cell isn't empty, it just can't be parsed at all.

Reply:
I see, so the idea is basically that if there's a major error such that we can't parse anything we will surface the error to the user. At first I thought the errors would be ignored downstream but now I traced it all the way back to the socket session client stream - so I can expect to see completion errors if I do something weird in my cell. 

Do you know what sort of errors would end up here? I don't recall seeing any completion errors while messing around with polynote .